### PR TITLE
CT-2280 offender sars validation

### PR DESCRIPTION
**Pull request by Polsonby**
## Description
Performing the correct minimal validation on each screen of the create case flow for Offender SAR

This is the reason why we have the form object; it does the work of displaying each step of the wizard and knows which validations are required on each step. We have all the validations in the Case::SAR::Offender model, but we have monkey patched ApplicationRecord with a `valid_attributes?` method which takes a hash of params to validate using the model's validators as opposed to calling `valid?` which validates all the attributes in one go. By definition we need to only validate certain fields at each step. 

Because of the way radio button groups work, nothing is sent in the params if neither button is selected, which defeats the partial validation which wants to validate only fields that are passed on each form submission. Accordingly, and rather annoyingly, we need to merge in some empty params if the radio group is empty.
 
## Self-review checklist

* [x] (1) Quick stakeholder demo done
* [ ] (2) Bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [x] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers
* [x] (7) Table indexes added if required
 
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2280

### Screenshots
validations latest

### CODE REVIEW

File: app/form_models/offender_sar_case_form.rb [hunk 1]
```diff
@@ -27,10 +28,12 @@ class OffenderSARCaseForm
            :received_date_dd,
            :received_date_mm,
            :received_date_yyyy,
+           :received_date,
            :reply_method,
            :creator,
            :send_by_post?,
            :send_by_email?,
+           :foo,
```
Comment (mseedat-moj) [inline app/form_models/offender_sar_case_form.rb]: Accidental?

File: app/form_models/offender_sar_case_form.rb [hunk 2]
```diff
@@ -72,14 +75,40 @@ def valid?
...
+      params.delete("third_party_relationship") unless params["third_party"] == "true"
+
+      params.merge!("reply_method": "") unless params["reply_method"].present?
+      params.delete("email") unless params["reply_method"] == "send_by_email"
+      params.delete("postal_address") unless params["reply_method"] == "send_by_post"
+    when "date-received"
+      params.merge!("received_date" => "") unless params["received_date"].present?
+    end
+
+    params
```
Comment (mseedat-moj) [inline app/form_models/offender_sar_case_form.rb]: Is there a cleaner way of doing this by having say `defaults` and merging that with params if possible?

File: app/form_models/offender_sar_case_form.rb [hunk 3]
```diff
@@ -72,14 +75,40 @@ def valid?
...
+  end
+
   def build_case_from_session
-    values = @session[:offender_sar_state] || {}
+    values = @session[:offender_sar_state] || {date_of_birth: nil}
```
Comment (mseedat-moj) [inline app/form_models/offender_sar_case_form.rb]: Space in braces `{ date_of_birth: nil }`

File: app/models/case/sar/offender.rb [hunk 4]
```diff
@@ -1,33 +1,72 @@
-class Case::SAR::Offender < Case::SAR::Standard
+class Case::SAR::Offender < Case::Base
   class << self
     def type_abbreviation
       'OFFENDER'
```
Comment (mseedat-moj) [inline app/models/case/sar/offender.rb]: Can this be changed to `OFFENDER_SAR`

File: app/models/case/sar/offender.rb [hunk 5]
```diff
@@ -1,33 +1,72 @@
...
+                 previous_case_numbers: :string,
+                 other_subject_ids: :string,
+                 external_deadline: :date,
+                 date_of_birth: :date,
+                 third_party: :boolean,
+                 third_party_relationship: :string,
+                 subject_type: :string,
+                 received_date: :date,
+                 reply_method: :string,
+                 flag_for_disclosure_specialists: :boolean
```
Comment (mseedat-moj) [inline app/models/case/sar/offender.rb]: Can the list of attributes be ordered alphabetically and do any of them require default values?

File: spec/form_models/offender_sar_case_form_spec.rb [hunk 6]
```diff
@@ -63,23 +63,25 @@
 
   describe "#valid_attributes?" do
     context "when params is empty" do
-      it "returns true" do
-        expect(case_form.valid_attributes?({})).to be true
+      params = ActionController::Parameters.new( {} ).permit!
```
Comment (mseedat-moj) [inline spec/form_models/offender_sar_case_form_spec.rb]: Extra spaces around braces?

**Review by mseedat-moj — CHANGES_REQUESTED** [review]
Looks on the right track. Some initial cursory comments

File: app/form_models/offender_sar_case_form.rb (hunk 1 above)
Comment (Polsonby) [reply app/form_models/offender_sar_case_form.rb]: Indeed! Fixed, thanks :)

File: app/form_models/offender_sar_case_form.rb (hunk 3 above)
Comment (Polsonby) [reply app/form_models/offender_sar_case_form.rb]: Fixed thanks

File: app/models/case/sar/offender.rb (hunk 4 above)
Comment (Polsonby) [reply app/models/case/sar/offender.rb]: I've had a go at this and it breaks things; obviously I'm not changing it in all the right places...! But I see you're changing this in your PR and I think we should not change it here therefore

File: spec/form_models/offender_sar_case_form_spec.rb (hunk 6 above)
Comment (Polsonby) [reply spec/form_models/offender_sar_case_form_spec.rb]: Fixed, thanks!

File: app/models/case/sar/offender.rb (hunk 5 above)
Comment (Polsonby) [reply app/models/case/sar/offender.rb]: Changed, and no, I don't think so...?

File: app/form_models/offender_sar_case_form.rb (hunk 2 above)
Comment (Polsonby) [reply app/form_models/offender_sar_case_form.rb]: Thanks for the suggestion! I agree it was really hard to read previously - have had a go at using methods to clarify what's going on and added a comment, is this any better?

**Review by mseedat-moj — APPROVED** [review]
Looks good thanks for the changes

**Pull request merged**